packages/appium/lib/appium.js

```diff
@@ -241,6 +252,330 @@ class AppiumDriver extends DriverCore {
     }
   }
 
+  /**
+   * Initialize a new bidi connection and set up handlers
+   * @param {import('ws').WebSocket} ws The websocket connection object
+   * @param {import('http').IncomingMessage} req The connection pathname, which might include the session id
+   */
+  onBidiConnection(ws, req) {
+    // wrap all of the handler logic with exception handling so if something blows up we can log
+    // and close the websocket
+    try {
+      const {bidiHandlerDriver, proxyClient, send, sendToProxy, logSocketErr} = this.initBidiSocket(
+        ws,
+        req,
+      );
+
+      this.initBidiSocketHandlers(
+        ws,
+        proxyClient,
+        send,
+        sendToProxy,
+        bidiHandlerDriver,
+        logSocketErr,
+      );
+      this.initBidiProxyHandlers(proxyClient, ws, send);
+      this.initBidiEventListeners(ws, bidiHandlerDriver, send);
+    } catch (err) {
+      this.log.error(err);
+      try {
+        ws.close();
+      } catch (ign) {}
+    }
+  }
+
+  /**
+   * Initialize a new bidi connection
+   * @param {import('ws').WebSocket} ws The websocket connection object
+   * @param {import('http').IncomingMessage} req The connection pathname, which might include the session id
+   */
+  initBidiSocket(ws, req) {
+    let outOfBandErrorPrefix = '';
+    const pathname = req.url;
+    if (!pathname) {
+      throw new Error('Invalid connection request');
```

Comment:
I would prefer to have a more specific error message, for example: `Invalid connection request: request pathname is missing`